### ISSUE TITLE
Mongo selects

### ISSUE THREAD

**Issue by jhiver** [TorchlightSoftware/mongo-watch#2]
Could you please explain how to use mongo selectors? I tried this:

  now = new Date()
  watcher.query {collName: 'archive', select: { ts: { $gt: now } } }, (err, query) ->

So that mongo-watcher wouldn't tail changes which happened in the past, but I got this:

MongoError: Unsupported projection option: $gt

Thanks in advance for your insight!

**Comment by bitmage** [comment]
My apologies... the API on this project was redesigned to serve the needs of a larger project, and it's a little rough to use it by hand.  If you give me some details on what you're trying to do, I might be able to point you towards some good resources.

As to the mechanics... it looks like there's a misunderstanding on the use of 'select'.  This is only for field selection, not for limiting the set of records.  It's akin to a SQL 'select' statement, which Mongo terms 'project': http://docs.mongodb.org/manual/reference/aggregation/project/

Mongo-Watch only supports limiting the recordset by ID.  So you can do this:

``` coffee-script
now = new Date()

# do a regular Mongo lookup to find out what IDs are present
getArchiveIdsLaterThan now, (err, idSet) ->
  watcher.query {collName: 'archive', idSet: idSet }, (err, query) ->
```

If you're wondering why we can't support a regular query, its because the delta stream we're tapping into is only guaranteed to include the _id field, so we can't filter on anything else.  The parent project, Particle, exposes a nicer API and does more work around query translation.  https://github.com/torchlightsoftware/particle

Hope that helps!

**Comment by jhiver** [comment]
Thanks for the quick reply... let me explain what I'm trying to do:

I'm using the mongo-watcher to watch for changes on a collection. However when I run it, it pulls lots of changes from the past which I don't need and or don't want to see.

While it's trivial to fix by doing the following within watcher.query:

```
  return if Number(event.data.created) < Number(startupTime)
```

I'm not happy about mongo-watcher eating around 2 Gb of memory until I can get anything useful done (this is what's happening on my system, with around 115000 Records being pulled from the oplog).

Do you have a solution for this?

Cheers
JM

**Comment by bitmage** [comment]
Well, long story short, we've got a lot of functionality here that doesn't belong in Mongo-Watch.

I think what I need to do is change Mongo-Watch back to its previous functionality, which simply tapped into the oplog.  It sounds like that would be more useful for you (and probably others), and I can move the complex query stuff into Particle.

Check this out:

https://github.com/TorchlightSoftware/mongo-watch/tree/df2bf8e65fbe0ddb97480ace1778669fafc47e22

If that looks like it will do your bidding (feel free to git clone it and test directly), then I will update the NPM version to point to that.

**Comment by jhiver** [comment]
Actually I think it's quite good that you're getting a replication of the data... although a "tail" method would be nice.

I see two places in your code where you use the toArray() method... I think this line is creating havoc:

./QueryPayload.coffee:      collection.find(@query, @options.select).toArray (err, results) =>

Because you're slurping everything into a huge list, then passing this huge list to formatPayload.

I've tried changing your code to:

```
  cursor = collection.find(@query, @options.select) # .toArray (err, results) =>
  cursor.each (err, doc) =>
    return @emit 'error', err if err
    return @push t: 'ep', op: 'n', ns: 'test.users', ts: getTimestamp() unless doc
    @push formatPayload [ doc ], @options
```

But I get the following error, any ideas what I'm doing wrong?

TypeError: Cannot read property 'high_' of undefined
  at module.exports.util.getDate (/Users/Mimi/Code/bushido/node_modules/mongo-watch/lib/util.coffee:63:32)
  at QueryStream.module.exports.pretty [as formatter](/Users/Mimi/Code/bushido/node_modules/mongo-watch/lib/formats.coffee:28:20)
  at QueryStream._transform (/Users/Mimi/Code/bushido/node_modules/mongo-watch/lib/QueryStream.coffee:185:21)
  at QueryStream.Transform._read (_stream_transform.js:179:10)
  at QueryStream.Transform._write (_stream_transform.js:167:12)
  at doWrite (_stream_writable.js:219:10)
  at writeOrBuffer (_stream_writable.js:209:5)
  at QueryStream.Writable.write (_stream_writable.js:180:11)
  at write (_stream_readable.js:573:24)
  at flow (_stream_readable.js:582:7)
  at QueryPayload.pipeOnReadable (_stream_readable.js:614:5)
  at QueryPayload.EventEmitter.emit (events.js:92:17)
  at emitReadable_ (_stream_readable.js:408:10)
  at emitReadable (_stream_readable.js:404:5)
  at readableAddChunk (_stream_readable.js:165:9)
  at QueryPayload.Readable.push (_stream_readable.js:127:10)
  at /Users/Mimi/Code/bushido/node_modules/mongo-watch/lib/QueryPayload.coffee:76:24
  at /Users/Mimi/Code/bushido/node_modules/mongodb/lib/mongodb/cursor.js:244:9
  at Cursor.nextObject (/Users/Mimi/Code/bushido/node_modules/mongodb/lib/mongodb/cursor.js:678:5)
  at commandHandler (/Users/Mimi/Code/bushido/node_modules/mongodb/lib/mongodb/cursor.js:658:14)
  at /Users/Mimi/Code/bushido/node_modules/mongodb/lib/mongodb/db.js:1670:9
  at Server.Base._callHandler (/Users/Mimi/Code/bushido/node_modules/mongodb/lib/mongodb/connection/base.js:383:41)
  at /Users/Mimi/Code/bushido/node_modules/mongodb/lib/mongodb/connection/server.js:472:18
  at [object Object].MongoReply.parseBody (/Users/Mimi/Code/bushido/node_modules/mongodb/lib/mongodb/responses/mongo_reply.js:68:5)
  at [object Object].<anonymous> (/Users/Mimi/Code/bushido/node_modules/mongodb/lib/mongodb/connection/server.js:430:20)
  at [object Object].EventEmitter.emit (events.js:95:17)
  at [object Object].<anonymous> (/Users/Mimi/Code/bushido/node_modules/mongodb/lib/mongodb/connection/connection_pool.js:191:13)
  at [object Object].EventEmitter.emit (events.js:98:17)
  at Socket.<anonymous> (/Users/Mimi/Code/bushido/node_modules/mongodb/lib/mongodb/connection/connection.js:301:18)
  at Socket.EventEmitter.emit (events.js:95:17)
  at Socket.<anonymous> (_stream_readable.js:736:14)
  at Socket.EventEmitter.emit (events.js:92:17)
  at emitReadable_ (_stream_readable.js:408:10)
  at emitReadable (_stream_readable.js:404:5)
  at readableAddChunk (_stream_readable.js:165:9)
  at Socket.Readable.push (_stream_readable.js:127:10)
  at TCP.onread (net.js:526:21)

**Comment by bitmage** [comment]
Hmm... can you explain what it is you expect mongo-watch to do, and how this ties into your application objectives?

You should _not_ be getting a huge data set from this.  If you're still using the current API, you probably need to do an idSet limit.  If you want to use the old API (as suggested above) it will only give you changes that occur from that point on.

**Comment by jhiver** [comment]
Hi,

Firstly, I'm expecting mongo-watch NOT to slurp so much memory to start with. I think it could be great if you avoided using toArray() and used cursor's each() method instead.

Secondly, I think it would be great if there were two methods:

1 / sync() - what the new API does, i.e. send the entire collection + the diffs, this way the process using this method can create an up to date set or subset of the collection you're watching

2/ tail() - what the old API does, i.e. "just the diffs from now on"

Let me know if you think that makes sense and if I can help with it in any way.

I actually managed to grab some of that code you wrote and turn it into something that's useful for me - thanks a lot for this BTW - but I think your library has so much more potential...

Cheers
JM

**Comment by bitmage** [comment]
I think the present version is better in terms of performance.  It's been a while since I made the changes, but if I recall there was an update to Mongo itself that makes tailing the oplog more efficient, and the method I'm using now works well with it.

Regarding potential and getting back into the data synchronization problem... you're right JM, lots of potential there.  It's a problem that interests me, but unfortunately my focus has had to shift to other areas.

I'm also leaning away from the oplog as a basis for events...  The design goals of the oplog are often at odds with what we really need, which is a 'triggers' system.  In lieu of Mongo actually supporting triggers, I think that a listener at the ORM level pushing to a message bus would be a firmer foundation than listening to the oplog.

Somewhat recent example: Around Mongo 2.21 or so, they decided to switch the oplog format from recording an array addition as just the single item being added, to recording the entire new contents of the array.  This was to achieve idempotency, an important quality for the oplog sync process.  But this threw a wrench in my plans, because I wanted to know the delta, both for performance reasons and because the UI updates I was making were based on a 'concat', not a 'update'.  You'll never be notified of changes like these, and they are likely to occur on minor versions, because this is considered an internal mechanism of Mongo, and was not really intended to be coupled with outside systems.

Long story short, I've largely abandoned mongo-watch for my purposes.  I'll keep this simplified version of the library around, but if I had to solve this problem again I would use a message bus instead.

**Comment by jhiver** [comment]
Hello Brandon,

I totally understand, and I agree with you, a trigger system is really
what's needed.

Thanks for that library of yours, it taught me a lot!

Cheers
JM
